File: homeworks/homework2/22002105_Gokmen_AhmetBerke.py
# Are formal and actial parameters type checked?
def fun(x, y, z):
  return x + y + z


print(fun("1", "2", '3'))
print(fun(1, 2, 3))


# Are keyword (named) parameters supported?
def kw_params(x, y, z):
  return x * z - y


print(kw_params(y=10, z=30, x=40))
print(kw_params(20, z=50, y=60))


# Are default paremeters supported?
def default_params(x, y, z=10):
  return x * y / z


print(default_params(10, 12))
print(default_params(y=10, x=12))

# What are the paremeter passing methods provided?
def default_params(a, b, c):  
	a = 10
	b.append(10)
	c += 30
	print("in", a, b, c)

# The real way's name is pass by assignment
a = 15  # treated as pass by value
b = [1] # treated as pass by reference
c = 30 # treated as pass by value
default_params(a, b, c)
print("out", a, b, c)


# Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?
text = "outside"

def subprogram_param(fun):
  text = "inside"
  return fun()


def sub():
  return text


print(subprogram_param(sub))

File: homeworks/homework2/22002105_Gokmen_AhmetBerke.rb
# 1. Are formal and actial parameters type checked?
puts "\n", "1."

def type_check(var1, var2, var3)
   puts var1, var2, var3
end

type_check "sa", 2, 3
type_check 4, 10.4, "as"

# 2. Are keyword (named) parameters supported?
puts "\n", "2."

def kw_args(name:, surname:)
  puts name, surname
end

kw_args name: "berke", surname: "gokmen"
kw_args surname: "gokmen", name: "berke"

# 3. Are default paremeters supported?
puts "\n", "3."

def default_param (var1 = "sa", var2: 10)
   puts var1, var2
end

default_param "as", var2: 30
default_param "as"
default_param

# 4. What are the paremeter passing methods provided?
puts "\n", "4."

# Ruby uses pass by object reference (python terminology)
def pass_param(obj)
  obj["x"] = "sa" # this will change the original object
end

def pass_number(num) 
  num += 30 # this will not affect the original object
end

number = 10
param = {"x" => 123}
puts param
puts number
pass_param(param) # this changes the original object
pass_number(number) # this does not change the original object
puts param
puts number

# 5. Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?
puts "\n", "5."

# the referencing environment of the passed subprogram is it's own
def perform_operation(operation)
	num = 200
  operation.call()
end

num = 99

def print_num()
  num = 100
  puts "inside", num # this will use 100
end

# This is the way of passing methods as parameters in ruby it seems
perform_operation(method(:print_num))